### PR TITLE
[NCL-7640] Add support for upgrade paths for deprecated environments

### PR DESCRIPTION
**Pull request by janinko**
Add new endpoint to deprecate environments
Set DEPRECATION_UPGRADE attribute when an environment is being deprecated.

### Checklist:

* [ ] Have you added a note in the [CHANGELOG wiki](https://github.com/project-ncl/pnc/wiki/Changelog) for your change if user-facing?
* [x] Have you added unit tests for your change?

### CODE REVIEW

**Comment by janinko** [comment]
Requires https://github.com/project-ncl/pnc-api/pull/138 to be merged

File: facade/src/main/java/org/jboss/pnc/facade/providers/EnvironmentProviderImpl.java [hunk 1]
```diff
@@ -48,7 +51,9 @@ public EnvironmentProviderImpl(BuildEnvironmentRepository repository, Environmen
     @Override
     @RolesAllowed(SYSTEM_USER)
     public Environment store(Environment restEntity) throws DTOValidationException {
-        return super.store(restEntity);
+        Environment storedEntity = super.store(restEntity);
+        deprecateActiveEnvironmentsWithSameName(storedEntity);
+        return storedEntity;
```
Comment (pkocandr) [inline facade/src/main/java/org/jboss/pnc/facade/providers/EnvironmentProviderImpl.java:56]: I'm missing one thing here. When we deprecate an older version of an env which was an upgrade target for another deprecated env, should we update also the DEPRECATION_UPGRADE value which pointed to the currently deprecated environment to point at the new one? That way we would return the upgrade target right away and save the user from having to walk through the deprecated path iteratively.

Comment (pkocandr) [reply facade/src/main/java/org/jboss/pnc/facade/providers/EnvironmentProviderImpl.java:56]: Or maybe not here, but in the deprecateActiveEnvironmentsWithSameName().

File: spi/src/main/java/org/jboss/pnc/spi/datastore/predicates/EnvironmentPredicates.java [hunk 2]
```diff
@@ -17,10 +17,15 @@
...
 package org.jboss.pnc.spi.datastore.predicates;
 
+import org.jboss.pnc.api.constants.Attributes;
 import org.jboss.pnc.model.BuildEnvironment;
 import org.jboss.pnc.model.BuildEnvironment_;
 import org.jboss.pnc.spi.datastore.repositories.api.Predicate;
 
+import javax.persistence.criteria.Expression;
+import javax.persistence.criteria.MapJoin;
+import java.util.Map;
```
Comment (pkocandr) [inline spi/src/main/java/org/jboss/pnc/spi/datastore/predicates/EnvironmentPredicates.java:27]: Another unused import.

**Review by pkocandr — CHANGES_REQUESTED** [review]
I believe the updated envs need to be saved explicitely. The unused imports are just a cherry on top :-)

File: facade/src/main/java/org/jboss/pnc/facade/providers/EnvironmentProviderImpl.java [hunk 3]
```diff
@@ -58,21 +64,54 @@ public void delete(String id) {
...
+        }
+        deprecatedEnvironment.setDeprecated(true);
+        deprecatedEnvironment.putAttribute(Attributes.DEPRECATION_REPLACEMENT, replacementId);
+        shortenTransitivePath(id, replacementId);
+        return mapper.toDTO(deprecatedEnvironment);
+    }
+
+    private void shortenTransitivePath(String deprecatedId, String replacementId) {
+        for (BuildEnvironment environment : repository.queryWithPredicates(replacedBy(deprecatedId))) {
+            environment.putAttribute(Attributes.DEPRECATION_REPLACEMENT, replacementId);
```
[the remark this answers is not in the document]
Comment (pkocandr) [reply facade/src/main/java/org/jboss/pnc/facade/providers/EnvironmentProviderImpl.java:113]: OK.

**Review by pkocandr — APPROVED** [review]

**Pull request merged**